src/modules/ekf2/ekf2_main.cpp

```diff
@@ -947,39 +949,42 @@ void Ekf2::run()
 
 		// get external vision data
 		// if error estimates are unavailable, use parameter defined defaults
-		bool vision_position_updated = false;
-		bool vision_attitude_updated = false;
-		orb_check(_ev_pos_sub, &vision_position_updated);
-		orb_check(_ev_att_sub, &vision_attitude_updated);
-
-		if (vision_position_updated || vision_attitude_updated) {
-			// copy both attitude & position if either updated, we need both to fill a single ext_vision_message
-			vehicle_attitude_s ev_att = {};
-			orb_copy(ORB_ID(vehicle_vision_attitude), _ev_att_sub, &ev_att);
+		bool visual_odometry_updated = false;
+		orb_check(_ev_odom_sub, &visual_odometry_updated);
 
-			vehicle_local_position_s ev_pos = {};
-			orb_copy(ORB_ID(vehicle_vision_position), _ev_pos_sub, &ev_pos);
+		if (visual_odometry_updated) {
+			// copy odometry data to fill a single ext_vision_message
+			vehicle_local_position_s ev_odom;
+			orb_copy(ORB_ID(vehicle_visual_odometry), _ev_odom_sub, &ev_odom);
 
 			ext_vision_message ev_data;
-			ev_data.posNED(0) = ev_pos.x;
-			ev_data.posNED(1) = ev_pos.y;
-			ev_data.posNED(2) = ev_pos.z;
-			matrix::Quatf q(ev_att.q);
-			ev_data.quat = q;
+			ev_data.posNED(0) = ev_odom.x;
+			ev_data.posNED(1) = ev_odom.y;
+			ev_data.posNED(2) = ev_odom.z;
+			ev_data.quat = matrix::Quatf(ev_odom.q);
 
 			// position measurement error from parameters. TODO : use covariances from topic
-			ev_data.posErr = fmaxf(_ev_pos_noise.get(), fmaxf(ev_pos.eph, ev_pos.epv));
-			ev_data.angErr = _ev_ang_noise.get();
+			ev_data.posErr = fmaxf(_ev_pos_noise.get(), fmaxf(ev_odom.eph, ev_odom.epv));
+			ev_data.angErr = fmaxf(_ev_pos_noise.get(), ev_odom.att_std_dev);
+
+			ev_odom.xy_valid = (!PX4_ISNAN(ev_odom.eph) && ev_odom.eph > ep_max_std_dev) ? false : true;
+			ev_odom.z_valid = (!PX4_ISNAN(ev_odom.epv) && ev_odom.epv > ep_max_std_dev) ? false : true;
+			ev_odom.v_xy_valid = (!PX4_ISNAN(ev_odom.evh) && ev_odom.evh > ev_max_std_dev) ? false : true;
+			ev_odom.v_z_valid = (!PX4_ISNAN(ev_odom.evv) && ev_odom.evv > ev_max_std_dev) ? false : true;
+
+			// only set data if the pose is valid
+			bool att_valid = false;
+
+			if (!PX4_ISNAN(ev_odom.att_std_dev)) {
+				att_valid = ev_odom.att_std_dev < eo_max_std_dev;
+			}
 
-			// only set data if all positions and velocities are valid
-			if (ev_pos.xy_valid && ev_pos.z_valid && ev_pos.v_xy_valid && ev_pos.v_z_valid) {
+			if (ev_odom.xy_valid && ev_odom.z_valid && att_valid) {
```

Comment:
There's not really any point in populating the incoming struct like this, in fact you should probably be treating it like const.

Reply:
Couldn't an RTPS client be conceivable setting them directly?

Reply:
My idea is to have bitmask instead.

> Couldn't an RTPS client be conceivable setting them directly?

We shouldn't mix the subjects: the idea behind this message restructure is not have a specific pipeline for RTPS, but rather something that could be used by both RTPS and Mavlink parser module.

Reply:
I'm talking about not using fields from an incoming message as temp storage. Ignore these fields for vehicle_visual_odometry if it's appropriate, but don't write into them out of convenience. It adds unnecessary confusion for someone reading or extending this later on.